service/worker/indexer/esProcessor.go

```diff
@@ -92,17 +92,64 @@ func newESProcessor(
 		return nil, err
 	}
 
-	p.bulkProcessor = processor
+	p.bulkProcessor = []bulk.GenericBulkProcessor{processor}
+	p.mapToKafkaMsg = collection.NewShardedConcurrentTxMap(1024, p.hashFn)
+	return p, nil
+}
+
+// newESDualProcessor creates new ESDualProcessor which handles double writes to dual visibility clients
+func newESDualProcessor(
+	name string,
+	config *Config,
+	esclient es.GenericClient,
+	osclient es.GenericClient,
+	logger log.Logger,
+	metricsClient metrics.Client,
+) (*ESProcessorImpl, error) {
+	p := &ESProcessorImpl{
+		config:     config,
+		logger:     logger.WithTags(tag.ComponentIndexerESProcessor),
+		scope:      metricsClient.Scope(metrics.ESProcessorScope),
+		msgEncoder: defaultEncoder,
+	}
+
+	params := &bulk.BulkProcessorParameters{
+		Name:          name,
+		NumOfWorkers:  config.ESProcessorNumOfWorkers(),
+		BulkActions:   config.ESProcessorBulkActions(),
+		BulkSize:      config.ESProcessorBulkSize(),
+		FlushInterval: config.ESProcessorFlushInterval(),
+		Backoff:       bulk.NewExponentialBackoff(esProcessorInitialRetryInterval, esProcessorMaxRetryInterval),
+		BeforeFunc:    p.bulkBeforeAction,
+		AfterFunc:     p.bulkAfterAction,
+	}
+	esprocessor, err := esclient.RunBulkProcessor(context.Background(), params)
+	if err != nil {
+		return nil, err
+	}
+
+	params.AfterFunc = p.shadowBulkAfterAction
+	osprocessor, err := osclient.RunBulkProcessor(context.Background(), params)
```

Comment:
nit: Could you add a comment here? It's not easy to understand for people not familiar with the code.